mypy/test/test_check.py

```diff
@@ -122,13 +122,12 @@ def run_test_once(self, testcase: DataDrivenTestCase, incremental=0) -> None:
             a = e.messages
         a = normalize_error_messages(a)
 
-        if output != a and mypy.myunit.UPDATE_TESTCASES:
+        if output != a and pytest.config.getoption('UPDATE_TESTCASES'):
             update_testcase_output(testcase, a, mypy.myunit.APPEND_TESTCASES)
 
-        assert_string_arrays_equal(
-            output, a,
+        assert output == a, \
             'Invalid type checker output ({}, line {})'.format(
-                testcase.file, testcase.line))
+                testcase.file, testcase.line)
```

Comment:
Yeah, I've used pytest in other projects, and I appreciate the fancy output on asserts. The thing is that `assert_string_arrays_equal` is really kind of a misnomer for the purpose that that helper serves for us in our myunit tests. It's not about just arrays of strings -- it's really optimized for lists of *lines*. And it's quite good at that, in a way that the generic pytest handling of lists isn't for this case.

For example, say I mess up one of the `check` tests, `testListAssignmentFromTuple`, like so:
```
 [a, c], c = t, c  # E: Incompatible types in assignment (expression has type "B", variable has type "C")
-[a, a, a], c = t, c  # E: Need more than 2 values to unpack (3 expected)
+[a, a, a], c = t, c  # E: Need more than 2 values to unpack (4 expected)
 [a], c = t, c  # E: Too many values to unpack (1 expected, 2 provided)
```

Our existing myunit framework will print this output:
```
$ PYTHONPATH=$PWD scripts/myunit -m mypy.test.testcheck '*ListAssignmentFromTuple'
Expected:
  main:6: error: Incompatible types in assignment (expression has type "B", v...
  main:7: error: Need more than 2 values to unpack (4 expected) (diff)
  main:8: error: Too many values to unpack (1 expected, 2 provided)
Actual:
  main:6: error: Incompatible types in assignment (expression has type "B", v...
  main:7: error: Need more than 2 values to unpack (3 expected) (diff)
  main:8: error: Too many values to unpack (1 expected, 2 provided)

Alignment of first line difference:
  E: ...2 values to unpack (4 expected)
  A: ...2 values to unpack (3 expected)
                            ^
Traceback (most recent call last):
  File "/home/greg/w/mypy/mypy/test/data.py", line 144, in run
    self.perform(self)
  File "/home/greg/w/mypy/mypy/test/testcheck.py", line 93, in run_test
    self.run_test_once(testcase)
  File "/home/greg/w/mypy/mypy/test/testcheck.py", line 144, in run_test_once
    testcase.file, testcase.line))
  File "/home/greg/w/mypy/mypy/test/helpers.py", line 85, in assert_string_arrays_equal
    raise AssertionFailure(msg)
AssertionFailure: Invalid type checker output (/home/greg/w/mypy/test-data/unit/check-lists.test, line 34)

test_testcheck_TypeCheckSuite.testListAssignmentFromTuple failed

1/1 test cases failed.
*** FAILURE ***
```

It prints both the actual and expected arrays as a full line for each line -- so that they look nearly just like the actual way these error messages would come out in normal operation. (They're slightly indented, which is probably helpful, and also abbreviated, which maybe we actually shouldn't do.) That's extremely useful when debugging a failure.

Then it also highlights the mismatched lines with the suffix " (diff)", and it even drills into the first pair of different lines and highlights the column where they first differ.

Your current version will print this:
$ time py.test -k ListAssignmentFromTuple mypy/test/test_check.py ============================ test session starts =============================
platform linux -- Python 3.5.1, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: /home/greg/w/mypy, inifile: pytest.ini
collected 1386 items 

mypy/test/test_check.py F

================================== FAILURES ==================================
_________________ TestTypeCheck.testListAssignmentFromTuple __________________

self = <mypy.test.test_check.TestTypeCheck object at 0xb64874cc>
test = <mypy.test.data.DataDrivenTestCase object at 0xb668c4ac>

    def func(self, test):
>       test.run(self)

../../mypy/test/helpers.py:314: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../mypy/test/data.py:163: in run
    self.perform(obj, self)
../../mypy/test/test_check.py:82: in run_test
    self.run_test_once(testcase)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <mypy.test.test_check.TestTypeCheck object at 0xb64874cc>
testcase = <mypy.test.data.DataDrivenTestCase object at 0xb668c4ac>
incremental = 0

    def run_test_once(self, testcase: DataDrivenTestCase, incremental=0) -> None:
        find_module_clear_caches()
        pyversion = pyversion_testcase(testcase.file, testcase.name)
        program_text = '\n'.join(testcase.input)
        module_name, program_name, program_text = self.parse_options(program_text)
        flags = self.parse_flags(program_text)
        output = testcase.output
        if incremental:
            flags.append(build.INCREMENTAL)
            if incremental == 1:
                # In run 1, copy program text to program file.
                output = []
                with open(program_name, 'w') as f:
                    f.write(program_text)
                    program_text = None
            elif incremental == 2:
                # In run 2, copy *.py.next files to *.py files.
                for dn, dirs, files in os.walk(os.curdir):
                    for file in files:
                        if file.endswith('.py.next'):
                            full = os.path.join(dn, file)
                            target = full[:-5]
                            shutil.copy(full, target)
        source = BuildSource(program_name, module_name, program_text)
        try:
            res = build.build(target=build.TYPE_CHECK,
                              sources=[source],
                              pyversion=pyversion,
                              flags=flags + [build.TEST_BUILTINS],
                              alt_lib_path=test_temp_dir)
            a = res.errors
        except CompileError as e:
            res = None
            a = e.messages
        a = normalize_error_messages(a)
    
        if output != a and pytest.config.getoption('UPDATE_TESTCASES'):
            update_testcase_output(testcase, a, mypy.myunit.APPEND_TESTCASES)
    
>       assert output == a, \
            'Invalid type checker output ({}, line {})'.format(
                testcase.file, testcase.line)
E       AssertionError: Invalid type checker output (mypy/test/data/check-lists.test, line 34)
E       assert ['main:6: err... 2 provided)'] == ['main:6: erro... 2 provided)']
E         At index 1 diff: 'main:7: error: Need more than 2 values to unpack (4 expected)' != 'main:7: error: Need more than 2 values to unpack (3 expected)'
E         Use -v to get the full diff

../../mypy/test/test_check.py:128: AssertionError
============ 1385 tests deselected by '-kListAssignmentFromTuple' ============
================= 1 failed, 1385 deselected in 2.56 seconds ==================
```

It's a lot harder here to tell what's going on. Partly that's because there's a lot of goo which pytest is printing out that's from the test framework and from the `test_check` test driver -- none of which will be at issue for >95% of the times that someone is looking at a test failure -- rather than from this particular test case. That's annoying but I think it's OK to have in an initial version of this; we can try later to cut it down.

More critically, there just isn't some of the useful information that's there in the myunit version. The differing pair of lines themselves are printed in full, but stuck together on one giant line with extra stuff around them, because the format is intended for shorter strings or other values. And the context lines aren't shown at all.

I can add `-v` as the message says, and then I get this at the end:
```
>       assert output == a, \
            'Invalid type checker output ({}, line {})'.format(
                testcase.file, testcase.line)
E       AssertionError: Invalid type checker output (mypy/test/data/check-lists.test, line 34)
E       assert ['main:6: err... 2 provided)'] == ['main:6: erro... 2 provided)']
E         At index 1 diff: 'main:7: error: Need more than 2 values to unpack (4 expected)' != 'main:7: error: Need more than 2 values to unpack (3 expected)'
E         Full diff:
E         ['main:6: error: Incompatible types in assignment (expression has type "B", '
E         'variable has type "C")',
E         -  'main:7: error: Need more than 2 values to unpack (4 expected)',
E         ?                                                     ^
E         +  'main:7: error: Need more than 2 values to unpack (3 expected)',
E         ?                                                     ^
E         'main:8: error: Too many values to unpack (1 expected, 2 provided)']
```
That's better but still I think a good deal harder to read than the myunit version. And it's an extra step.

So I'd like to see this PR leave `assert_string_arrays_equal` in the loop and continue to show the kind of output it produces, one way or another. There are surely improvements to make to that helper, and perhaps even ways to integrate it usefully with pytest fanciness, but those can be separate PRs to keep the scope of this one from growing.